### PR TITLE
Make internal and external types equivalent

### PR DESCRIPTION
**Pull request by gagern**
As discussed in #409, these changes turn the separate set of opaque types into equally opaque pointers to incomplete structures (i.e. declarations without definitions), with the complete definition added internally. In other words, the public API still doesn't know about the internal structure of any of these objects, but the tidy code can transparently operate them without any need to convert back and forth. As a consequence, the code allows more robust type-checking, is shorter and easier to read. Currently this diff removes 194 lines. There probably is some room for further improvements, since many of the public API functions are now just aliases for some private function, so that in a later commit we could eliminate the private names and instead implement the public functions directly. But the diff is long enough as it is, so I'll not do that just now.

`TidyIterator` doesn't actually name a single structure.  Often it's not a pointer at all, but an integer instead. So that will still be dealt with as before.

The `TidyOption` definition is somewhat special as well, since it contains the const qualifier, and the name of the structure doesn't follow the common schema. For that reason it doesn't make use of the `opaque_type` macro.

### CODE REVIEW

**Comment by geoffmcl** [comment]
@gagern sorry, but do not understand? 

You say "As discussed in #409"? What was discussed? Anything agreed? Did I miss something?

As to the PR, so far all I can see is a lot of find and replace... ok, to what gain exactly? 

How will this improve tidy? What **bug** does it address? What is the desired `features` added? What do you find difficult to read in the code, where this will make it easier?, etc, etc, etc... ?

And just as an aside, please target such a large proposed code change in a branch, maybe `issue-409`, or something... but **not** the `master` branch... this has to be tested, agreed, discussed, understood, ... before committing it to the `master` branch...

But it looks like you have an **idea**, but can not yet quite see what it is ;=)) thanks, I think...

**Comment by gagern** [comment]
In #409 I suggested what I did here, and your reply didn't indicate any pressing reason against this, so I went forward and coded the changes to make further discussion more specific.

This does not address an existing bug. It does not add a new feature. It is about maintainability. In my opinion, the change addresses the following issues:

1. Converting between external and internal types adds noise: additional function calls, additional variables, for what's essentially a no-op at the level of compiled code. 194 lines less code means 194 lines less to understand and maintain, so things become simpler.
2. Having different names for external and internal types can make you forget how they relate, while having the same names throughout helps users who know the public API to become contributors who enhance the internal API.
3. The type conversion macros did cast to the target type, without verifying the source type. Therefore, the compiler wouldn't issue a warning if you happened to pass in an inappropriate pointer into one of these conversion macros. (The conversion functions don't have this problem, but those were `NEVER` used.)
4. I'm no expert on this, but the fact that the opaque structures had some dummy definition might make them run afoul of some [strict pointer aliasing rules](https://gcc.gnu.org/onlinedocs/gcc-6.1.0/gcc/Optimize-Options.html#index-fstrict-aliasing-872). While I can't think of a piece of code where actual aliasing would occur, the compiler makes no guarantees so we might end up with undefined behavior. This would be a real bug, so perhaps there is a bug to be fixed after all.
5. Many other projects use pointers to incomplete structures in their public API. The fact that Tidy does something more elaborate made me (and may make others) wonder as to the reason for that added complexity, while apparently there is none.
6. Having equal types means that it's easier to promote a previously internal function to public API without having to change its signature.

I *did* write this in a separate branch. Namely [the one from which I'm proposing the merge](https://github.com/gagern/tidy-html5/commits/opaque). Feel free to test it from there. Feel free to discuss it here. If you need help setting up an additional git remote, I'll be glad to provide details. If you want things in the htacg repository, you'll have to duplicate the branch there, since I can only issue pull requests for existing branches.

I don't want you to merge this branch just because you trust me. Test and discuss as you see fit, ask questions, suggest changes. When you are satisfied, and convinced that this adds value, then merge this pull request. Since the change touches large portions of the code, I'll likely have to rebase it before you can merge it. But I'd like to do so only once, when you think it's ready for merging. I won't track every master commit here.

Most of the changes are indeed search and replace. And search and delete in the case of the conversion macro invocations. So let me direct your attention to the less trivial aspects.

1. The `opaque_type` macro in `tidyplatform.h` is now without definition, leading to an incomplete declaration. You can have pointers of those, but no other access to the structures.
2. The corresponding definitions in `lexer.h` and `tidy-int.h` now add a definition to the structure already declared in `tidy.h`.
3. In `forward.h` the declarations for the internal structures `TidyDocImpl`, `AttVal` and `Node` are gone.
4. The conversion macros in `tidy-int.h` and conversion functions in `tidylib.c` are gone.

That's the actual core of the change, the rest is just a consequence of that.

**Comment by gagern** [comment]
I found [some useful information on the strict aliasing rules](http://stackoverflow.com/a/7291895/1468366). According to that, neither the old nor the new style of handling opaque types should be a problem, as long as noone attempts to access the (dummy) contents of the opaque type (i.e. dereferences the pointer). The only possible problem might come from alignment constraints.

On the other hand, some of the `TidyIterator` handling is likely relying on undefined behavior because conversion between integers and pointers isn't officially supported. But my change here doesn't improve that aspect; that would be a different issue.

**Comment by geoffmcl** [comment]
@gagern, have downloaded, built, and done some testing of the result... which I versioned `5.3.1-exposed` :=))

While this PR has the diffs, a total diff I am reviewing is here - http:/geoffair.org/tmp/gagern.diff

And then paused a few weeks, waiting for comments from others... ok, none at this point... just you and me...

To be right up front, I do not like the merge yet! There are some **good** points, but to me it fails to reach the `50%` plus swaying point... really sorry...

Just to make sure we are on the same page, what I see is essentially what you have done is made a `TidyDoc doc` usable in `external` and `internal` code! That seems to defeat some of the idea of a tidy defined **opaque** type, but ok, I understand you are offering an alternative, that also works somewhat simularly? 

And done the same with some other pointers, like the `node *`, attr *... and I think added another... I do understand the different approaches to the same idea...

Just to put it in, hopefully, simple terms, currently an external app is passed a `TidyDoc tdoc` pointer. The fact that it is `tidy defined` opaque pointer means it can not address the structure, and equally important can **not** be use to call an internal library functions, not in the API, without most compilers complaining... aim achieved in current code...

When that `opaque` pointer is passed into the library it is **converted** to a `TidyDcoImp *`, which all/most internal libery services/functions use... Same with several other pointers...

After a fundimental header definition change, this leads to a mostly cut-and-paste, almost cosmetic, difference, like -

````
-void TY_(AccessibilityChecks)( TidyDocImpl* doc );
+void TY_(AccessibilityChecks)( TidyDoc doc );
````

So if we **agree** that is the discussion here, then to address your 6 points...

 1. Yes, if your calculations are correct we could save a few hundred instructions, which are just `stack a = b;`... But I do not think it really effects readbility to an exterienced programmer, or code maintainability overall... No.

 2. It could help new programmers, in that it is a tiny tidy code learning step. But once known, understood, is no problem... Yes.

 3. If a coder happened to pass in an inappropriate pointer into the API, they should suffer the consequence of that error, which will not always be flagged by every compiler/linker, **what ever we do**, so can not see how this improves that... No.

 4. I too know of no bug this fixes... sort of change with what gained? but open...

 5. How other projects do this is their business... the tidy creators constructed their own maco... why are we now choosing to abandon this? for simplicity? If it is wrong fix it, otherwise leave it alone... open

 6. ... easier to promote an internal service to the API? Yes, since with this change all internal library functions become very close to public, just not yet in `tidy.h`, ie not published... open...

Then you point us to the less trivial, fundimental changes in the important `definitions`

 1. `tidyplatform.h` yes, you changed the macro to a simpler macro, and changed the comment more to your choice!
 2. `lexer.h` and `tidy-int.h` yes, we understand, one change means more...
 3. `forward.h` as above, consequential to the first change...
 4. `tidy-int.h` more consequence to the above choice...

So the choice remains - should we stick with the current implemetation, or switch a few important definitions, and then accept 2. 3. 4.? A very difficult yes/no choice, at least for me... open...
 
So, in summary, while I agree there is some reduced verbosity and noise on pointers, I do not yet see a real gain at this time, and again wonder why we should change something that was put there by the tidy founders, that works... 

So to the central question here - Make internal and external types equivalent - I would suggest **no**!

But will certainly wait for other comments... thanks for the effort, and all the work done here... that is very great! Shows a fundimental understanding of tidy code... 

I hope you will apply the same enthusiam and efforts to fixing some other tidy **issues**...

**Comment by gagern** [comment]
> […] which I versioned `5.3.1-exposed`

I'd consider that a misnomer, since it's not exposing any more than before.

> While this PR has the diffs, a total diff I am reviewing is here - [http:/geoffair.org/tmp/gagern.diff](http:/geoffair.org/tmp/gagern.diff)

That diff appears to be equivalent to what we have here in the “Files changed” tab, apart from end-of-line convention and path separator character following Windows. So I don't see the point of the distinction.

> Just to make sure we are on the same page, what I see is essentially what you have done is made a `TidyDoc doc` usable in `external` and `internal` code!

Yes.

> That seems to defeat some of the idea of a tidy defined opaque type, but ok, I understand you are offering an alternative, that also works somewhat simularly?

The data structures are opaque to the outside and transparent to the inside, both before and after my change. Previous code achieves the distinction by type casts, my code by incomplete types.

> Just to put it in, hopefully, simple terms […]. So if we **agree** that is the discussion here […]

I think we agree.


1. >> Converting between external and internal types adds noise: additional function calls, additional variables, for what's essentially a no-op at the level of compiled code. 194 lines less code means 194 lines less to understand and maintain, so things become simpler.

   > Yes, if your calculations are correct we could save a few hundred instructions, which are just `stack a = b;`... But I do not think it really effects readbility to an exterienced programmer, or code maintainability overall... No.

  I agree that performance is not an issue at this stage. Perhaps even the assignment you mention could be optimized away by compilers.

  Things might be a bit different if we use this simplification to remove redundant code in a future commit. For example, the public `tidySaveFile` is now (with the casts gone away) a simple forwarding to the private `tidyDocSaveFile`. We could just use the name of the former and the code of the latter to get rid of one function to maintain, and one function call on the call stack and performance bill. And there are several such forwarding functions. If you consider these savings relevant, I can add a commit to get rid of these as well.

2. >> Having different names for external and internal types can make you forget how they relate, while having the same names throughout helps users who know the public API to become contributors who enhance the internal API.

   > It could help new programmers, in that it is a tiny tidy code learning step. But once known, understood, is no problem... Yes.

3. >> The type conversion macros did cast to the target type, without verifying the source type. Therefore, the compiler wouldn't issue a warning if you happened to pass in an inappropriate pointer into one of these conversion macros. (The conversion functions don't have this problem, but those were `NEVER` used.)

   > If a coder happened to pass in an inappropriate pointer into the API, they should suffer the consequence of that error, which will not always be flagged by every compiler/linker, **what ever we do**, so can not see how this improves that... No.

 It's not an external user passing code to our API, but more about internal code. You are of course right that using wrong pointers should lead to broken runtime behavior, and that properly tested code should catch those in time. But personally I like the fact that a C compiler can detect many trivial mistakes before I even start testing code. Perhaps a matter of taste. Although seeing the state of affairs with respect to the tidyhtml test suite, I'd consider any help we can get pointing out mistakes to be a valuable one.

4. Strict aliasing should not be an issue, as I wrote.

5. >> Many other projects use pointers to incomplete structures in their public API. The fact that Tidy does something more elaborate made me (and may make others) wonder as to the reason for that added complexity, while apparently there is none.

   > How other projects do this is their business... the tidy creators constructed their own maco... why are we now choosing to abandon this? for simplicity? If it is wrong fix it, otherwise leave it alone... open

 For simplicity, yes. To lower the burden for potential contributors. With my strict aliasing concerns dispelled, the existing code is not wrong, just not very idiomatic either.

6. >> Having equal types means that it's easier to promote a previously internal function to public API without having to change its signature.

   > ... easier to promote an internal service to the API? Yes, since with this change all internal library functions become very close to public, just not yet in `tidy.h`, ie not published... open...

 Right. Which doesn't mean one should make functions public without due consideration, but after that consideration things become easier.

> I […] wonder why we should change something that was put there by the tidy founders, that works...

I wonder why the founders wrote it that way in the first place. It seems as though they didn't like to declare structures without defining them. Even the dummy structures used for the current opaque types come with a bogus definition. Perhaps there were some compilers not capable of properly handling such constructs back when the project started? I don't know, but I'm rather sure that these concepts are well-supported these days, and have been for quite some time.

> I hope you will apply the same enthusiam and efforts to fixing some other tidy **issues**...

I tend to address issues that affect me. I wanted to tidy HTML in Node, so I started the [node-libtidy project](https://github.com/gagern/node-libtidy). I see that it has some problems with how arguments are handled in error reporting, and would like to address those concerns by introducing a new opaque type. And I'd have preferred to do so my way, i.e. using an incomplete type in the public API. But I guess I'll rather stay with the style currently in use, and not wait on this issue here since I don't see this here getting accepted any time soon.

**Comment by geoffmcl** [comment]
@gagern I do not think there would be any objection to a **new** opaque type being added, using the style currently in use...

And just a small negative to the suggested change here, when using the MSVC **Visual** C/C++ IDE, this opacity is very clear in the IDE local variable display, when debugging...

Presently the the tidy `opaque` pointer and the `impl` pointer show up **visually** different, making the situation very clean -

````
- tdoc	0x00000000004e64f0 {_opaque=0x00000000 }	const _TidyDoc *
	_opaque	0x00000000	int
- impl	0x00000000004e64f0 {root={...} lexer=0x0000000000000000 config={...} ...}	_TidyDocImpl *
 + 	root	{parent=0x0000000000000000 prev=0x0000000000000000 next=0x0000000000000000 ...}	_Node
 +	lexer	0x0000000000000000 {lines=??? columns=??? waswhite=??? ...}	_Lexer *
 + ... etc
````

That seems good, and very clear...

In the proposed `5.3.1-exposed` version there are not two pointers, they are the same, so the IDE will show the `doc` in full,  all the time - in and outside the library code - like -

````
- doc	0x00000000004e64f0 {root={...} lexer=0x0000000000000000 config={...} ...}	_TidyDocImpl *
 + 	root	{parent=0x0000000000000000 prev=0x0000000000000000 next=0x0000000000000000 ...}	_Node
 +	lexer	0x0000000000000000 {lines=??? columns=??? waswhite=??? ...}	_Lexer *
 + ... etc
````

So there is is a loss of a visual clue to the `opacity` of the first, versus the real `impl` structure pointer... making it **look** like you have access to say `doc->config.value[0].v`... of course the compiler will still belch an error if outside the library...

So of course the compile guards are still there. If you try to write code like this outside the library, the IDE/compiler will quickly stop you... but it does now **visually** imply you could...

Maybe not so important to coders that do not have or use a **visual** IDE to write code... but MSVC coders get to reply on it...

Anyway, I think you have correctly assessed the situation - without further support, reasons, ... - this merge will not be made, and I hope we can move on... thanks...

**Comment by geoffmcl** [comment]
In the absence of further supporting comments, am closing this for the reasons given... but thanks for the work involved... it certainly shows another way to handle `transparent` types...
